From 036bf692ea8c4e821a8af8bff8304e683c06ffa4 Mon Sep 17 00:00:00 2001
From: Muhammad Falak R Wani <falakreyaz@gmail.com>
Date: Sat, 2 Sep 2023 10:28:37 +0530
Subject: [PATCH 3/5] Clear scrollback buffer when locking terminal
    
Linux kernel 3.0 adds a new command to clear the scrollback buffer in
the linux console. It's available in the ncurses terminfo entry as
extended capability E3. This is useful to prevent an attacker to read
content of a locked terminal via Shift-PgUp key.


Signed-off-by: Muhammad Falak R Wani <falakreyaz@gmail.com>
---
 display.c | 6 ++++++
 extern.h  | 1 +
 screen.c  | 1 +
 term.c    | 1 +
 4 files changed, 9 insertions(+)

diff --git a/display.c b/display.c
index a1afbe6..1c56b3e 100644
--- a/display.c
+++ b/display.c
@@ -4080,4 +4080,10 @@ char **cmdv;
 
 #endif /* BLANKER_PRG */
 
+void
+ClearScrollbackBuffer()
+{
+  if (D_CE3)
+    AddCStr(D_CE3);
+}
 
diff --git a/extern.h b/extern.h
index d4e57df..4564581 100644
--- a/extern.h
+++ b/extern.h
@@ -320,6 +320,7 @@ extern int   color256to88 __P((int));
 extern void  ResetIdle __P((void));
 extern void  KillBlanker __P((void));
 extern void  DisplaySleep1000 __P((int, int));
+extern void  ClearScrollbackBuffer __P((void));
 
 /* resize.c */
 extern int   ChangeWindowSize __P((struct win *, int, int, int));
diff --git a/screen.c b/screen.c
index 7653cd1..85d52ac 100644
--- a/screen.c
+++ b/screen.c
@@ -1970,6 +1970,7 @@ void Detach(int mode)
 
     case D_LOCK:
       ClearAll();
+      ClearScrollbackBuffer();
       sign = SIG_LOCK;
       /* tell attacher to lock terminal with a lockprg. */
       break;
diff --git a/term.c b/term.c
index c7a13c9..f49b3a7 100644
--- a/term.c
+++ b/term.c
@@ -83,6 +83,7 @@ struct term term[T_N] =
   { "CD", T_STR  },
   { "ce", T_STR  },
   { "cb", T_STR  },
+  { "E3", T_STR  },
 
 /* initialise */
   { "is", T_STR  },
-- 
2.40.1

